### PR TITLE
Support Command.output_paths

### PR DESCRIPTION
**Pull request by iain-macdonald**
This PR does a few things:
- Adds a more complex test for the various kinds of outputs (output_directories, output_files, and output_paths)
- Slightly changes the treatment of output_directories to match the description in the protocol here: https://github.com/bazelbuild/remote-apis/blob/main/build/bazel/remote/execution/v2/remote_execution.proto#L619 Specifically, before this change the executor would create all of the specified output_directories whereas now it will only create the parents of all of the specified output_directories. The behavior for output_files is unchanged.
- Uses output_paths instead of output_directories and output_files if specified, otherwise falls back to output_directories and output_files.

**Version bump**: Minor

Related issues: Fixes https://github.com/buildbuddy-io/buildbuddy-internal/issues/1766

### CODE REVIEW

**Comment by iain-macdonald** [comment]
FYI I haven't yet managed to test this locally, still working on that. I'll let you know when I have.

**Comment by iain-macdonald** [comment]
I managed to test this locally by changing this logic in Bazel: https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java#L233  I'll check with Siggi today to see if it works with his setup as well.

File: dockerfiles/executor_image/Dockerfile [hunk 1]
```diff
@@ -19,7 +19,7 @@ RUN apt-get update && apt-get install -y umoci iproute2
 # Install podman and skopeo
 RUN echo 'deb https://downloadcontent.opensuse.org/repositories/home:/alvistack/Debian_11/ /' >> /etc/apt/sources.list.d/home:alvistack.list && \
     curl -fsSL https://download.opensuse.org/repositories/home:/alvistack/Debian_11/Release.key | gpg --dearmor >> /etc/apt/trusted.gpg.d/home_alvistack_debian11.gpg && \
-    apt-get update && apt-get -y upgrade && apt-get install -y podman=100:4.2.1-1 skopeo=100:1.9.2-1
+    apt-get update && apt-get -y upgrade && apt-get install -y podman=100:4.3.0-1 skopeo=100:1.10.0-1
```
[the remark this answers is not in the document]
Comment (iain-macdonald) [reply dockerfiles/executor_image/Dockerfile]: Whoops, nope, gone now.

File: enterprise/server/remote_execution/dirtools/dirtools.go [hunk 2]
```diff
@@ -118,39 +107,44 @@ func (c *DirHelper) CreateOutputDirs() error {
 	}
 	return nil
 }
-func (c *DirHelper) MatchesOutputDir(path string) (string, bool) {
-	for _, d := range c.outputDirs {
-		for p := path; p != filepath.Dir(p); p = filepath.Dir(p) {
-			if p == d {
-				return d, true
-			}
+func (c *DirHelper) IsOutputPath(path string) bool {
```
[the remark this answers is not in the document]
Comment (iain-macdonald) [reply enterprise/server/remote_execution/dirtools/dirtools.go:112]: Yup, done

File: enterprise/server/test/integration/remote_execution/remote_execution_test.go [hunk 3]
```diff
@@ -787,23 +789,23 @@ func TestBasicActionIO(t *testing.T) {
 		Arguments: []string{
 			"sh", "-c", strings.Join([]string{
 				`set -e`,
+				// Make the output directory -- even though it's declared in
```
[the remark this answers is not in the document]
Comment (iain-macdonald) [reply enterprise/server/test/integration/remote_execution/remote_execution_test.go:792]: Done

[the remark this answers is not in the document]
Comment (iain-macdonald) [reply enterprise/server/test/integration/remote_execution/remote_execution_test.go:792]: Looks like there is a tool that can do this automatically: https://github.com/segmentio/golines  I tried it and it works okay... support for comments is not so great. We could consider adding it to buildfix.

File: enterprise/server/remote_execution/dirtools/dirtools.go [hunk 4]
```diff
@@ -118,39 +107,44 @@ func (c *DirHelper) CreateOutputDirs() error {
...
+	_, ok := c.outputPaths[path]
+	return ok
+}
+
+// If the provided path is a descendant of any of the output_paths, this function returns which
+// output_path it is a descendent of and true, otherwise it returns an unspecified string and false.
```
[the remark this answers is not in the document]
Comment (iain-macdonald) [reply enterprise/server/remote_execution/dirtools/dirtools.go]: Done

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

**Review by tylerwilliams — APPROVED** [review]
LGTM

File: enterprise/server/test/integration/remote_execution/remote_execution_test.go [hunk 5]
```diff
@@ -940,6 +942,195 @@ func TestComplexActionIO(t *testing.T) {
...
+	outDir := rbe.DownloadOutputsToNewTempDir(res)
+
+	expectedFiles := []string{
+		"a/a/a.txt", "a/a/b.txt", "a/b/a.txt", "a/b/b.txt", "b/a/a.txt",
+        "b/a/b.txt", "c/a/a.txt", "c/a/b.txt", "c/b/a.txt", "d/a.txt",
+        "d/a/a.txt",
+	}
+	unexpectedFiles := []string{
+        "a.txt", "b.txt", "b/b", "c/b/b.txt", "d/a/b.txt", "d/b", "e",
+    }
```
[the remark this answers is not in the document]
Comment (iain-macdonald) [reply enterprise/server/test/integration/remote_execution/remote_execution_test.go]: Whoops, yep

**Pull request merged**